### ISSUE TITLE
Prometheus: Error on ingesting samples with different value but same timestamp

### ISSUE THREAD

**Issue by axi92** [NCCloud/harbor-vulnerabilities-exporter#9]
I upgraded prometheus from 2.51.2 to 2.55.1 and since then I get a lot of errors like this:
I had this with harbor-vulnerabilities-exporter:0.0.3 and harbor-vulnerabilities-exporter:0.0.4 now as well.

```
ts=2024-12-16T08:51:51.570Z caller=scrape.go:1820 level=warn component="scrape manager" scrape_pool=harbor-vulnerabilities target=http://10.64.191.113:8000/ msg="Error on ingesting samples with different value but same timestamp" num_dropped=238
```
This metric was enhanced in this prometheus release https://github.com/prometheus/prometheus/releases/tag/v2.52.0
PR: https://github.com/prometheus/prometheus/pull/12933

**Comment by vriabyk** [comment]
hello @axi92, thank you for posting the issue. I've double checked our environments, we also use `2.55.1` prometheus version deployed via [kube-prometheus-stack](https://artifacthub.io/packages/helm/prometheus-community/kube-prometheus-stack) helm chart and `harbor-vulnerabilities-exporter:0.0.4`. There are no such errors in our prometheus logs.

We need more details from you in order to help with the further investigation. 
* can you please dump metrics to the file and send it to us here? something like `curl http://10.64.191.113:8000/metrics > metrics.txt` or if you will run it inside prometheus pod - it will probably won't have `curl`, but `wget` should be available.
* any other errors in `prometheus` or `harbor-vulnerabilities-exporter` logs?
* if possible share details/values how you deploy the exporter in your envs.

**Comment by axi92** [comment]
I use this metric to trigger an alert rule `increase(prometheus_target_scrapes_sample_duplicate_timestamp_total[5m]) > 0`
And it looks like this in prometheus:
![image](https://github.com/user-attachments/assets/2e1b67a0-0049-4ba2-b808-4379ee1a7a5a)

I noticed that the lines are not every time the same.
With this little helper script I tried to find the rare requests when I got more or less lines:
```
#!/bin/bash
RAW=$(curl http://10.64.191.113:8000/)
LINES=$(echo "${RAW}" | wc -l)
DATE=$(date +"%Y-%m-%dT%H:%M:%S%z")
echo ${LINES}
echo "${RAW}" > ${LINES}_${DATE}.log
```
The output is:
```
165K Dec 16 13:11 214_2024-12-16T13:11:51+0100.log
300K Dec 16 13:05 373_2024-12-16T13:05:18+0100.log
300K Dec 16 13:05 373_2024-12-16T13:05:20+0100.log
300K Dec 16 13:05 373_2024-12-16T13:05:23+0100.log
300K Dec 16 13:05 373_2024-12-16T13:05:25+0100.log
300K Dec 16 13:05 373_2024-12-16T13:05:36+0100.log
300K Dec 16 13:05 373_2024-12-16T13:05:38+0100.log
300K Dec 16 13:05 373_2024-12-16T13:05:39+0100.log
300K Dec 16 13:05 373_2024-12-16T13:05:41+0100.log
300K Dec 16 13:05 373_2024-12-16T13:05:45+0100.log
300K Dec 16 13:11 373_2024-12-16T13:11:38+0100.log
300K Dec 16 13:11 373_2024-12-16T13:11:41+0100.log
300K Dec 16 13:11 373_2024-12-16T13:11:43+0100.log
300K Dec 16 13:11 373_2024-12-16T13:11:46+0100.log
300K Dec 16 13:11 373_2024-12-16T13:11:48+0100.log
300K Dec 16 13:11 373_2024-12-16T13:11:53+0100.log
300K Dec 16 13:11 373_2024-12-16T13:11:56+0100.log
300K Dec 16 13:11 373_2024-12-16T13:11:58+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:01+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:04+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:06+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:09+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:12+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:15+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:18+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:20+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:23+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:25+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:28+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:32+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:35+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:37+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:39+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:41+0100.log
300K Dec 16 13:12 373_2024-12-16T13:12:43+0100.log
434K Dec 16 13:05 532_2024-12-16T13:05:43+0100.log
```
So sometimes there are more lines and sometimes there are even less...
I put in the gist 3 files
You can see at the start of the filename the lines in the file.
The one with the 214 is missing even some repos.
373 is I guess the correct response because this is the response thats given 95% of the time I would say.
And then there is the 532 file with the duplicates.

Prometheus is just giving those warning all other logs are the typical info logs.

The deploymen is on an ubuntu vm just with a compose file:
```
  harbor-vulnerabilities-exporter:
    image: ghcr.io/nccloud/harbor-vulnerabilities-exporter:0.0.4
    container_name: harbor-vulnerabilities-exporter
    hostname: harbor-vulnerabilities-exporter
    restart: unless-stopped
    environment:
      - HARBOR_API_URL=${HARBOR_API_URL}
      - HARBOR_USERNAME=${HARBOR_USERNAME}
      - HARBOR_PASSWORD=${HARBOR_PASSWORD}
      - THREADS=5
      - EXPORTER_PORT=8000
      - IGNORE_REPOSITORIES=${IGNORE_REPOSITORIES}
    ports:
      - ${HARBOR_BIND}:8000
```

**Comment by crabique** [comment]
@axi92 how many [replicas](https://github.com/NCCloud/charts/blob/main/charts/harbor-vulnerabilities-exporter/values.yaml#L2) are you running for the exporter?

**Comment by axi92** [comment]
I only have one running.

**Comment by vriabyk** [comment]
I've checked `prometheus_target_scrapes_sample_duplicate_timestamp_total` metric at our envs - it is constantly 0. So it doesn't sound like problem related to prometheus upgrade to me, maybe it was just hidden for you before upgrading to 2.55. Sorry, but have more questions:

* is exporter running on the same node with harbor installation? do you have any proxy/webserver in front of harbor?
* as far as you don't run it in k8s, how do you discover the exporter in prometheus? do you use static scrape config? what scrape interval do you set?

If you are sure there is no second instance of the exporter running somewhere (it wasn't actually developed to have replicas) - then it might be problem with too low scrape interval setting comparing to actual response time of the exporter. So the prometheus may not receive previous response while generates another one.
What I would recommend to you is to try to measure a time of response from the exporter: for example using `time` cli tool with `curl` and to ensure scrape interval is higher than it.

**Comment by axi92** [comment]
No the exporter is running on a different node. There is a nginx as a reverse proxy in-front of it.
I register the service in consul, consul is used as the target catalog in prometheus.
```
- job_name: harbor-vulnerabilities
  scrape_interval: 60s
  scheme: http
  metrics_path: /
  consul_sd_configs:
  - server: 172.17.0.1:8500
    services: [harbor-vulnerabilities]
```

Scrape interval is 60s and the response time is under 1s
```
real    0m0.862s
user    0m0.006s
sys     0m0.008s
```

**Comment by vriabyk** [comment]
@axi92 can you please try to set THREADS=1 and check if the problem exists? also interesting how it will affect response time in your case.

**Comment by axi92** [comment]
Response time is now 3-4 seconds, which is fine.
I tried it now several times with 373 lines and then there are some anomalies with 627 or 664 lines. Looks like it is still the same also with only 1 thread.
I just realized that it is requesting all the data every time from harbor when there is a request on harbor-vulnerabilities-exporter and after all the data is gathered the response is sent. Wouldn't it be less load if the exporter had a scheduler that gets the data every 10 or so minutes and buffers it for an instant response? I don't know the use-case why the data would be changing so often on harbor that the new data needs to be collected every time again?

**Comment by vriabyk** [comment]
> I just realized that it is requesting all the data every time from harbor when there is a request on harbor-vulnerabilities-exporter

yes, this is how prometheus exporters work usually. That's why there is scrape_interval setting in prometheus server which you can set per job level to control how often you wanna get metrics (trigger gathering from the source). If you have ideas how to improve it - PRs are always welcome!

I would suggest to set scrape_interval to 5 min at least.

I am checking files you provided above, will let you know if something is found.

**Comment by axi92** [comment]
Ah ok I understand. I tried that in the past but ran into staleness problems https://prometheus.io/docs/prometheus/latest/querying/basics/#staleness I did not touch that problem in a long time. Is it possible by now to set the stale setting for one target now? I don't want to set it for the whole prometheus instance.
So per default everything I think longer than 5min as scrape_interval counts as stale.

**Comment by vriabyk** [comment]
I was able to reproduce the problem with abnormal number of metrics (duplicated ones) ONLY when running two or more requests simultaneously to the harbor-vulnerabilities-exporter/metrics endpoint. Also the problem exists for parallel requests from prometheus-server (on scrape time) and for example curl from cli. When you run curl at the same time prometheus scrapes the target it even fails target scraping in prometheus:
![image](https://github.com/user-attachments/assets/f1b4468c-0cc8-4e7c-8697-e5c6941bcf43)

But othen that that I can't reproduce your the problem. As I mentioned above, the exporter was developed to run 1 replica only. We may extend it in the future if there will be real use cases.

> Is it possible by now to set the stale setting for one target now?

as far as I know - there is no such possibility. We set scrape_interval to 5 mins in our envs and don't face any problems.

**Comment by vriabyk** [comment]
Forgot to add, that I will check what we can do from code perspective, so it won't duplicate metrics on simultaneous requests. Because it looks like a bug to me.

**Comment by axi92** [comment]
Thank you for investigating my issue =)
Feel no pressure to fix this in the next 3 Weeks. I will be back at work in the second week of January, so I can test it if you have a fix by then. If not its also fine. I muted the alert for now =)

**Comment by vriabyk** [comment]
this should be fixed in 0.0.5 version

**Comment by axi92** [comment]
I upgraded to 0.0.5 and the issue looks resolved. Thanks for the fix!
